### PR TITLE
readwrite module for interfacing with RDF

### PR DESCRIPTION
**Pull request by pedros**
Read and write graphs in [RDF](https://www.w3.org/RDF/) format. Requires [rdflib](https://github.com/RDFLib/rdflib). This module adds 8 new functions:
- [from_rgmlgraph](https://github.com/pedros/networkx/blob/readwrite/rdf/networkx/readwrite/rdf.py#L280)
- [to_rgmlgraph](https://github.com/pedros/networkx/blob/readwrite/rdf/networkx/readwrite/rdf.py#L342)
- [read_rgml](https://github.com/pedros/networkx/blob/readwrite/rdf/networkx/readwrite/rdf.py#L392)
- [write_rgml](https://github.com/pedros/networkx/blob/readwrite/rdf/networkx/readwrite/rdf.py#L424)
- [from_rdfgraph](https://github.com/pedros/networkx/blob/readwrite/rdf/networkx/readwrite/rdf.py#L456)
- [to_rdfgraph](https://github.com/pedros/networkx/blob/readwrite/rdf/networkx/readwrite/rdf.py#L542)
- [read_rdf](https://github.com/pedros/networkx/blob/readwrite/rdf/networkx/readwrite/rdf.py#L579)
- [write_rdf](https://github.com/pedros/networkx/blob/readwrite/rdf/networkx/readwrite/rdf.py#L608)

It can load arbitrary RDF graphs, or RGML-namespaced graphs. [RGML](https://www.cs.rpi.edu/research/groups/pb/punin/public_html/RGML/) is the RDF Graph Modeling Language, an RDF ontology for generic graphs.

For arbitrary RDF graphs, it can represent them as multiple directed labeled graphs as per the RDF specification, with terms that occur as subject/object and, later on, as a predicate, being reified multiple times. Since this is not optimal for connectivity analyses, it also recognizes that RDF graphs are, in fact, hypergraphs, and as such represents them as bipartite graphs where each node in one partition has 3 edges pointing to subject, predicate, and object nodes in the other partition.

For more information, see [Hayes, J. (2004). A graph model for RDF. Technische Universität Darmstadt](http://www.dcc.uchile.cl/~cgutierr/papers/rdfgraphmodel.pdf)

All serialization formats supported by `rdflib` are supported, currently:
- [RDF/XML](http://www.w3.org/TR/rdf-syntax-grammar/)
- [Notation 3](http://www.w3.org/TeamSubmission/n3/)
- [Turtle](http://www.w3.org/TeamSubmission/turtle/)
- [N-Triples](http://www.w3.org/TR/rdf-testcases/#ntriples)
- [Trix](https://www.hpl.hp.com/techreports/2004/HPL-2004-56)

### CODE REVIEW

**Comment by MridulS** [comment]
@pedros Thanks for the contribution! And sorry for the delay in review 😅

As of today it looks like rdflib maintains connectors to networkx https://github.com/RDFLib/rdflib/blob/9625ed0b432c9085e2d9dda1fd8acf707b9022ab/rdflib/extras/external_graph_libs.py#L72 so we don't need to add them here :)

The link to RGML seems broken, I think this is the current one https://www.cs.rpi.edu/~puninj/rgml.html . And from previous experience maintaining markup languages in networkx becomes a maintenance burden so we are trying to avoid adding new readwrite modules, especially the ones which don't seem to have robust support outside of networkx already. Let me know what you think. Thanks again!

**Comment by rossbar** [comment]
> As of today it looks like rdflib maintains connectors to networkx so we don't need to add them here :)

I agree - there's not a lot to gain by duplicating format-conversion functionality, and the RDF library seems like a more natural place for these to live. I will go ahead and close this - thanks all for the proposal & discussion!

**Comment by westurner** [comment]
@MridulS

- [x] To copy from rdflib to networkx:
https://github.com/RDFLib/rdflib/blob/9625ed0b432c9085e2d9dda1fd8acf707b9022ab/rdflib/extras/external_graph_libs.py#L72

```python
def rdflib_to_networkx_digraph(
    graph,
    calc_weights=True,
    edge_attrs=lambda s, p, o: {"triples": [(s, p, o)]},
    **kwds,
):

def rdflib_to_networkx_multidigraph(
    graph, edge_attrs=lambda s, p, o: {"key": p}, **kwds
):
```

- [ ] To copy from rdflib to networkx:

@pedros @rossbar Would this rdflib read/write code be best as a third-party module?

https://github.com/networkx/networkx/blob/main/setup.py:

- [ ] `entry_points` to declare a plugin namespace:
  - https://github.com/networkx/networkx/blob/f2b337f1a4e37aa65766dd1638dc9c09a97d2687/networkx/classes/backends.py#L12 :
  ```python
      entry_points={'networkx.plugins': 'sparse = networkx_plugin_sparse'}
   ```

**Comment by MridulS** [comment]
@westurner The plugins bits are currently setup to work for backend computation plugins, not readwrite modules. But this is something that indeed can be thought more about :)

For readwrite to arrow/Parquet, I think we can have a readwrite inside networkx too! (just my opinion) Arrow is a robust data format outside of networkx and if there is an efficient way of reading/writing into that I think that's a plus.

Now if someone comes up and implements algorithms on top of arrow data structures for graphs, that would be great :D. We would be able to directly latch into that as a backend.

**Comment by westurner** [comment]
RDF support would be worthwhile as a core "read write plugin" or as a third-party adapter with it's own integration tests that depend upon rdflib import IMHO.

These have C/C++-based `tests_require`™dependencies, too:

> Now if someone comes up and implements algorithms on top of arrow data structures for graphs, that would be great :D. We would be able to directly latch into that as a backend.

https://github.com/rapidsai/cugraph/blob/branch-23.02/readme_pages/algorithms.md

https://github.com/rapidsai/cugraph#apache-arrow-on-gpu-- >

> ### Apache Arrow on GPU
> The GPU version of Apache Arrow is a common API that enables efficient interchange of tabular data between processes running on the GPU. End-to-end computation on the GPU avoids unnecessary copying and converting of data off the GPU, reducing compute time and cost for high-performance analytics common in artificial intelligence workloads. As the name implies, cuDF uses the Apache Arrow columnar data format on the GPU. Currently, a subset of the features in Apache Arrow are supported

https://www.phoronix.com/news/Intel-oneAPI-2023 :

> Very interesting as part of this is Intel-owned Codeplay Software releasing oneAPI plug-ins for NVIDIA and AMD GPUs. This allows SYCL and oneAPI use atop NVIDIA's proprietary driver stack as well as AMD with ROCm.

- https://github.com/tensorflow/io/blob/master/tensorflow_io/arrow.md
-  https://github.com/tensorflow/io/blob/master/tensorflow_io/python/ops/arrow_dataset_ops.py

**Comment by MridulS** [comment]
> RDF support would be worthwhile as a core "read write plugin" or as a third-party adapter with it's own integration tests that depend upon rdflib import IMHO.

rdflib already has all the support for conversion b/w networkx and rdf, not sure what else we can/should add.


> about arrow on GPU

Well we support (soon) cugraph as a backend for networkx so that's good news.

But arrow is still columnar memory layout which doesn't really work that well with graph algorithms, so it's not that straight forward having arrow support for the graph data structure itself. Yes, it can work as a dumping ground for graph data but not something we can write code on top off (which is the more interesting thing to me).

**Comment by westurner** [comment]
https://arrow.apache.org/powered_by/ Ctrl-F "graph" doesn't appear to list e.g.  CuGraph, which is built on Apache Arrow.

Does pyarrow already support SparseTensors?
https://arrow.apache.org/docs/cpp/api/tensor.html#sparse-tensors https://arrow.apache.org/docs/format/Other.html#sparse-tensor

> rdflib already has all the support for conversion b/w networkx and rdf, not sure what else we can/should add.

>> - [x] To copy from rdflib to networkx:

>> - [ ] To copy from (edit) networkx to rdflib

**Comment by westurner** [comment]
FWIW, rdflib-hdt also includes support for RDF HDT Header Dictionary Triples; which IIUC this PR would make easier to readwrite from?

**Comment by westurner** [comment]
TLDR Model serialization: https://en.wikipedia.org/wiki/Serialization

- https://github.com/EthicalML/awesome-production-machine-learning#model-serialisation
- https://github.com/maximveksler/awesome-serialization#machine-learning
- ONNX

Just found this, which mentions SQLAlchemy and rdflib: https://github.com/unum-cloud/NetworkXum#project-structure
